axum-extra/src/extract/cached.rs

```diff
@@ -0,0 +1,232 @@
+/// }
+///
+/// struct CurrentUser { /* ... */ }
+///
+/// #[async_trait]
+/// impl<B> FromRequest<B> for CurrentUser
+/// where
+///     B: Send,
+/// {
+///     type Rejection = Response<BoxBody>;
+///
+///     async fn from_request(req: &mut RequestParts<B>) -> Result<Self, Self::Rejection> {
+///         // loading a `CurrentUser` requires first loading the `Session`
+///         //
+///         // by using `Cached<Session>` we avoid extracting the session more than
+///         // once, in case other extractors for the same request also loads the session
+///         let session: Session = Cached::<Session>::from_request(req)
+///             .await
+///             .map_err(|err| err.into_response().map(body::boxed))?
+///             .0;
+///
+///         // load user from session...
+///         # unimplemented!()
+///     }
+/// }
+///
+/// // handler that extracts the current user and the session
+/// //
+/// // the session will only be loaded once, even though `CurrentUser`
+/// // also loads it
+/// async fn handler(
+///     current_user: CurrentUser,
+///     // we have to use `Cached<Session>` here otherwise the
+///     // cached session would not be used
+///     Cached(session): Cached<Session>,
+/// ) {
+///     // ...
+/// }
+/// ```
+///
+/// [request extensions]: http::Extensions
+#[derive(Debug, Clone, Default)]
+pub struct Cached<T>(pub T);
+
+#[async_trait]
+impl<B, T> FromRequest<B> for Cached<T>
+where
+    B: Send,
+    T: FromRequest<B> + Clone + Send + Sync + 'static,
+{
+    type Rejection = CachedRejection<T::Rejection>;
+
+    async fn from_request(req: &mut RequestParts<B>) -> Result<Self, Self::Rejection> {
+        match Extension::<T>::from_request(req).await {
+            Ok(value) => Ok(Self(value.0)),
+            Err(ExtensionRejection::ExtensionsAlreadyExtracted(err)) => {
+                Err(CachedRejection::ExtensionsAlreadyExtracted(err))
+            }
+            Err(_) => {
+                let value = T::from_request(req).await.map_err(CachedRejection::Inner)?;
+
+                req.extensions_mut()
+                    .ok_or_else(|| {
+                        CachedRejection::ExtensionsAlreadyExtracted(
+                            ExtensionsAlreadyExtracted::default(),
+                        )
+                    })?
+                    .insert(value.clone());
+
+                Ok(Self(value))
+            }
+        }
+    }
+}
+
+impl<T> Deref for Cached<T> {
+    type Target = T;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl<T> DerefMut for Cached<T> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.0
+    }
+}
+
+/// Rejection used for [`Cached`].
+///
+/// Contains one variant for each way the [`Cached`] extractor can fail.
+#[derive(Debug)]
+#[non_exhaustive]
+pub enum CachedRejection<T> {
+    #[allow(missing_docs)]
+    ExtensionsAlreadyExtracted(ExtensionsAlreadyExtracted),
+    #[allow(missing_docs)]
+    Inner(T),
```

Comment:
I was confused there for a bit thinking the `T` on `CachedRejection` was the same thing as the `T` on `Cached`. Maybe the type definition would be clearer if it was, and this variant was `Inner(T::Rejection)`?

Reply:
Using the type of the extractor complicates the trait bounds for `IntoResponse for CachedRejection` and the others. I've changed the type param from `T` to `R` which is maybe a bit more clear.

Reply:
Yeah, I think this helps a lot actually 🙂